File: member-service/src/main/java/com/hedvig/memberservice/services/member/MemberService.java
package com.hedvig.memberservice.services.member;

import com.hedvig.external.authentication.dto.ZignSecAuthenticationMethod;
import com.hedvig.external.bankID.bankIdTypes.CollectResponse;
import com.hedvig.memberservice.commands.BankIdSignCommand;
import com.hedvig.memberservice.commands.ZignSecSignCommand;
import com.hedvig.memberservice.commands.models.ZignSecAuthenticationMarket;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class MemberService {

  final private CommandGateway commandGateway;

  public MemberService(CommandGateway commandGateway) {
    this.commandGateway = commandGateway;
  }

  public void bankIdSignComplete(final long memberId, @NonNull final CollectResponse collectResponse) {
    this.commandGateway.sendAndWait(
        new
            BankIdSignCommand(
            memberId, collectResponse.getOrderRef(), collectResponse.getCompletionData().getSignature(), collectResponse.getCompletionData().getOcspResponse(),
            collectResponse.getCompletionData().getUser().getPersonalNumber()));
  }

  public void signComplete(
    final long memberId,
    @NonNull final UUID referenceId,
    @NonNull final String peronalNumber,
    @NonNull final String providerJsonResponse,
    @NonNull final ZignSecAuthenticationMethod authenticationMethod) {
    this.commandGateway.sendAndWait(
      new
              ZignSecSignCommand(memberId, referenceId, peronalNumber, providerJsonResponse, ZignSecAuthenticationMarket.Companion.fromAuthenticationMethod(authenticationMethod))
    );
  }
}

File: member-service/src/main/java/com/hedvig/memberservice/services/member/MemberService.kt
package com.hedvig.memberservice.services.member

import com.hedvig.external.authentication.dto.ZignSecAuthenticationMethod
import com.hedvig.external.bankID.bankIdTypes.CollectResponse
import com.hedvig.memberservice.commands.SwedishBankIdSignCommand
import com.hedvig.memberservice.commands.ZignSecSignCommand
import com.hedvig.memberservice.commands.models.ZignSecAuthenticationMarket.Companion.fromAuthenticationMethod
import org.axonframework.commandhandling.gateway.CommandGateway
import org.springframework.lang.NonNull
import org.springframework.lang.Nullable
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class MemberService(private val commandGateway: CommandGateway) {
    fun bankIdSignComplete(memberId: Long, collectResponse: CollectResponse) {
        commandGateway.sendAndWait<Any>(
            SwedishBankIdSignCommand(
                memberId,
                collectResponse.orderRef,
                collectResponse.completionData.signature,
                collectResponse.completionData.ocspResponse,
                collectResponse.completionData.user.personalNumber
            ))
    }

    fun signComplete(
        memberId: Long,
        referenceId: UUID,
        personalNumber: String,
        providerJsonResponse: String,
        authenticationMethod: ZignSecAuthenticationMethod,
        firstName: String?,
        lastName: String?
    ) {
        commandGateway.sendAndWait<Any>(
            ZignSecSignCommand(
                memberId,
                referenceId,
                personalNumber,
                providerJsonResponse,
                fromAuthenticationMethod(authenticationMethod),
                firstName,
                lastName
            )
        )
    }
}
